### ISSUE TITLE
Feature request: Infinite resolution screenshot

### ISSUE THREAD

**Issue by hmaarrfk** [pygfx/pygfx#907]
It comes up from time to time, to be able to render a given scene at a resolution much much higher than what is supported by webgpu. The limit here seems to be between 8k x 8x to 16k x 16k.

This can be useful in the case that one crafts a beautiful montage (lets take for example `examples/validation/validate_points_markers.py`) and we want to really showcase the full details on a mega high resolution printout (say a mural).

With the ability to design things that are "pixel perfect"  it may be desirable to really showcase the exact rendering details of the pygfx pipeline
![image](https://github.com/user-attachments/assets/1946ac69-9732-463b-a453-c3acf0c6f27c)

along with all of the imperfections:
![image](https://github.com/user-attachments/assets/03f61456-cc67-446c-9e29-e59fdb9cb752)

The idea would be that pygfx could simulate what a 100k x 100k (10Gigapixel offscreen canvas) canvas might look like, and issue the repeated rendering commands to webgpu in order to render the whole thing.

The few concerns that come to mind:
1. Maintaining frustrum culling on GPU.
2. Dealing with "pixel perfect" interpolation at the seam of tiles. While not entirely related this is reminescent of https://github.com/pygfx/pygfx/issues/905
3. The ability to define the rendering canvas dimensions
4. The ability to provide an output buffer.
5. The ability to specify if we want `rgba` or `rgb`  output buffer.

Initial ideas that came up in a conversation with almar is that we could play around with the `rect` feature of
the
```
renderer.render
```
calls and "stitch" them together.

**Comment by Korijn** [comment]
I mean... Just define your scene and render it in tiles? That's how it's always done I think.

**Comment by hmaarrfk** [comment]
> I mean... Just define your scene and render it in tiles? That's how it's always done I think.

Yeah as I was writing this "feature request" I started drafting a demo implementation.

I think there are a few parameters to get right which makes this tricky, and potentially worthy of an example:

1. Getting the camera parameters right. plus, minus, and the right orientation.
2. Getting the camera "orientated" correct in 3D space. (I probably should get better at internalizing quaternions)
   * I feel like this is different for orthographic cameras and perspective cameras.
3. Getting shaders that work on NDC coordinates to work corrrectly. In taking the example I showed above, I hit the fact that the background shader is written directly in NDC, so it is effectively "tiled" as you render.
4. There are many "edge cases" in the tiling. For example your tile size might be `100 x 100`, but you want to render something that is `1159 x 2101` pixels.

I think the combination of these 4 makes this question "non trivial".

**Comment by hmaarrfk** [comment]
Gets you 90% of the way there for orthographic cameras, using the idea to just "rendering multiple times"

* Assumes you have an orthographic camera
* Shaders do not use NDC coordinates in their computation (e.g. background)
* Still unclear to me what happens at these boundary conditions

<details><summary>some sample code for the trivial case</summary>

```python
def offscreen_render_high_res(
    canvas, camera, scene,
    upscale=5,
    offscreen_side=1024,
):
    import imageio
    from copy import deepcopy
    from tqdm import tqdm
    camera_state = camera.get_state()
    if camera_state['fov'] != 0.0:
        raise ValueError("This is only implemented for orthographic cameras")

    # Lets keep it simple, and render to a square...
    offscreen_canvas = OffscreenCanvas(size=(offscreen_side, offscreen_side))
    # Some users might be using their own renderer....
    # let them shoot themselves in the foot...
    offscreen_renderer = renderer.__class__(
        offscreen_canvas,
        # Want to copy over many of the parameters from the renderer
        # to ensure we get as close to the same rendering as possible
        pixel_ratio=renderer.pixel_ratio,
        pixel_filter=renderer.pixel_filter,
        show_fps=False,  # this is not really meaningful, is it???
        blend_mode=renderer.blend_mode,
        sort_objects=renderer.sort_objects,
        enable_events=False,
        gamma_correction=renderer.gamma_correction,
        # Can we copy over the device?
    )

    offscreen_camera = camera.__class__()
    offscreen_camera_state = deepcopy(camera_state)
    # del offscreen_camera_state['position']
    offscreen_canvas.request_draw(lambda: offscreen_renderer.render(scene, offscreen_camera))

    logical_size = canvas.get_logical_size()
    logical_size = (logical_size[0] * upscale, logical_size[1] * upscale)
    offscreen_logical_size = offscreen_canvas.get_logical_size()
    tiles = (
        int((logical_size[0] + offscreen_logical_size[0] - 1) / offscreen_logical_size[0]),
        int((logical_size[1] + offscreen_logical_size[1] - 1) / offscreen_logical_size[1]),
    )
    final_image = np.zeros((tiles[1] * offscreen_side, tiles[0] * offscreen_side, 3), dtype='uint8')
    for i in tqdm(np.ndindex(tiles), miniters=0, mininterval=0, total=tiles[0] * tiles[1]):
        offscreen_side = max(
            (offscreen_logical_size[1] / (logical_size[1])) * camera_state['height'],
            (offscreen_logical_size[0] / (logical_size[0])) * camera_state['width'],
        )
        offscreen_camera_state['height'] = offscreen_side
        offscreen_camera_state['width'] = offscreen_side
        # Don't touch z???
        # offscreen_camera_state['position'][2] = camera_state['position'][2]

        offset_x = (i[0] - 0.5 * (tiles[0] - 1)) * offscreen_side
        offset_y = (i[1] - 0.5 * (tiles[1] - 1)) * offscreen_side

        # TODO: Take into account "rotation" and "reference_up"????

        offscreen_camera_state['position'][0] = camera_state['position'][0] + offset_x
        # GPU coordinates typically flip the y direction, so flip it here too in the offset
        offscreen_camera_state['position'][1] = camera_state['position'][1] - offset_y

        offscreen_camera.set_state(offscreen_camera_state)
        img = np.asarray(offscreen_canvas.draw())[..., :3]
        final_image[
            offscreen_side * i[1]:offscreen_side * (i[1] + 1),
            offscreen_side * i[0]:offscreen_side * (i[0] + 1)
        ] = img
        # imageio.v3.imwrite(f"test_offscreen_rect_{i[1]:03d}_{i[0]:03d}.png", img)

    imageio.v3.imwrite("final_image.png", final_image)
```
</details>

**Comment by hmaarrfk** [comment]
Shower thoughts got me to figure out how to get the NDC coordinates to match up correctly. However my naive implementation requires additions to the global state and modifications to the background shader.
https://github.com/hmaarrfk/pygfx/pull/10

I'm not too happy about this implementation since it "costs" more to the GPU with this additional feature due to the inclusion of 4 new matrices in stdinfo_uniform_type.

**Comment by almarklein** [comment]
I played with this a bit. I started with modifying the camera to prepend an ndc transform to the projection. But then I realized that the logical size must be adjusted to keep everyting sized correctly, so it felt better to move the API to the renderer. See #909

**Comment by hmaarrfk** [comment]
As I thought of the initial problem statement, I envisioned this simulating a 100k x 100k screen.

This means that lines drawn with 10 pixel thickness, should be rendered as 10 pixels on the final canvas.
A font that is 20 pixels high in screenspace, should resolve on the final 100k x 100k canvas as also being 20 pixels high.

I think this is the "simplest" case, but I understand that this might not cover all use cases. Some users might want to scale things differently.

Finally, I'm happy to cleanup the shader in https://github.com/hmaarrfk/pygfx/pull/8
it shows how points in the world space, can be combined with NDC coordinates and the viewport to create a cool effect where it feels like the text is sticky to the objects.

**Comment by hmaarrfk** [comment]
Thanks almar!
